### PR TITLE
[IMP] website_slides: duration of a course

### PR DESCRIPTION
**Pull request by pko-odoo**
The duration of a course is now computed. The duration is equal the sum of
the duration of all videos, plus the estimated time to read any PDF, plus the
duration encoded on the other slides. The other types of slide have a field to
encode the duration during its creation (not mandatory). Also, the computation
take only the slides that are not archive.

The format of the duration of a course on its tile is modified. It's now
displayed like this : "x min" or "y hours x min". The duration displayed is
also rounded to quarter of hour.

Task ID : 2007294

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_slides/models/slide_slide.py [hunk 1]
```diff
@@ -287,6 +311,8 @@ def create(self, values):
                 values.setdefault(key, value)
 
         slide = super(Slide, self).create(values)
+        if values.get('completion_time'):
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py]: Unnecessary for a computed field.

Another way would be to do it in an onchange instead of a computed field and allow values from front-end.

File: addons/website_slides/static/src/js/slides_upload.js [hunk 2]
```diff
@@ -371,6 +372,15 @@ var SlideUploadDialog = Dialog.extend({
...
 
+        if (currentType === 'video'){
+            if(!this.$('.duration').hasClass('o_hidden')){
+                this.$('.duration').addClass('o_hidden');
+            }
+        } else {
+            if(this.$('.duration').hasClass('o_hidden')){
+                this.$('.duration').removeClass('o_hidden');
+            }
+        }
```
Comment (dbeguin) [inline addons/website_slides/static/src/js/slides_upload.js]: could this be simplified like : ?? 
```
if (currentType === 'video' && !this.$('.duration').hasClass('o_hidden')){
    this.$('.duration').addClass('o_hidden');
} else {
    this.$('.duration').removeClass('o_hidden');
}
```

File: addons/website_slides/models/slide_slide.py [hunk 3]
```diff
@@ -108,7 +108,7 @@ class Slide(models.Model):
     category_id = fields.Many2one('slide.category', string="Category", domain="[('channel_id', '=', channel_id)]")
     tag_ids = fields.Many2many('slide.tag', 'rel_slide_tag', 'slide_id', 'tag_id', string='Tags')
     is_preview = fields.Boolean('Is Preview', default=False, help="The course is accessible by anyone : the users don't need to join the channel to access the content of the course.")
-    completion_time = fields.Float('# Hours', digits=(10, 4), compute="_compute_duration", store=True)
+    completion_time = fields.Float('Duration', digits=(10, 4), compute="_compute_duration", store=True)
```
Comment (dbeguin) [inline addons/website_slides/models/slide_slide.py]: I think this should be done in another commit (first one probably)

File: addons/website_slides/models/slide_channel.py [hunk 4]
```diff
@@ -22,13 +22,16 @@ class ChannelUsersRelation(models.Model):
...
+                 'partner_id', 'channel_id.slide_partner_ids.slide_id.is_published',
+                 'channel_id.slide_partner_ids.slide_id.active')
     def _compute_completion(self):
         read_group_res = self.env['slide.slide.partner'].sudo().read_group(
             ['&', '&', ('channel_id', 'in', self.mapped('channel_id').ids),
              ('partner_id', 'in', self.mapped('partner_id').ids),
              ('completed', '=', True),
-             ('slide_id.is_published', '=', True)],
+             ('slide_id.is_published', '=', True),
+             ('slide_id.active', '=', True)],
```
[the remark this answers is not in the document]
Comment (dbeguin) [reply addons/website_slides/models/slide_channel.py]: or put this directly in first commit ?

File: addons/web/static/lib/pdfjs/src/core/core.js [hunk 5]
```diff
@@ -93,7 +93,7 @@ var Page = (function PageClosure() {
       // "The crop, bleed, trim, and art boxes should not ordinarily
       // extend beyond the boundaries of the media box. If they do, they are
       // effectively reduced to their intersection with the media box."
-      cropBox = Util.intersect(cropBox, mediaBox);
+      cropBox = PDFJS.Util.intersect(cropBox, mediaBox);
```
Comment (awa-odoo) [inline addons/web/static/lib/pdfjs/src/core/core.js]: @jeh-odoo @pko-odoo 

I don't know which one of you made this change since you both worked on this PR, so I ping you both.

Please be careful with this, you're not supposed to change code in third party libraries (pdfjs in this case).
These libraries could be updated at any time and your code will simply be gone (and the bug re-introduced).

If you're interested in the actual fix for this problem, you can check: https://github.com/odoo/odoo/pull/35518

**Comment by pedrobaeza** [comment]
Coming to this from codetriage.com. It seems this is stalled with conflict files, so should it be closed?

**Comment by tde-banana-odoo** [comment]
It was done in another PR finally; I found notably https://github.com/odoo/odoo/commit/5269d8c27f9f5e7b2d0ab86db6423f6ffa227f40#diff-5adb5c220546c08dd205b16f24d44d8a .

I close this one.
